Add ID dApp tutorial (#1316)
Concordium/concordium.github.io: source/mainnet/tutorials/using-ID-in-dApps/zk-proofs-tutorial.rst

.. _zk-proofs-tutorial:

===================
ZK proofs structure
===================

> Now we'll explore how to implement zero-knowledge proof verification using Concordium's identity framework, focusing on how ZK statements are structured.

Author: I agree, will replace.

> Now we'll explore how to implement zero-knowledge proof verification using Concordium's identity framework, focusing on how ZK statements are structured.

Reviewer: ```suggestion
Now we'll explore how to generate and verify zero-knowledge proofs using Concordium's identity framework, focusing on how ZK statements are structured.
```

Defining ZK statements
----------------------

> First, let's look at the ZK statements defined in the constants file:

Reviewer: ```suggestion
First, let's look at the ZK statements defined in the ``constants`` file:
```

> .. code-block:: javascript
> 
>   import { AtomicStatementV2 } from "@concordium/web-sdk"; // From constants.ts

Reviewer: ```suggestion
.. code-block:: javascript
  // From constants.ts

  import { AtomicStatementV2 } from "@concordium/web-sdk"; 
```

  // Context string for signing messages and generating ZK proofs
  export const CONTEXT_STRING = "CONCORDIUM_COMPLIANT_REWARD_DISTRIBUTION_DAPP";

  // ZK statements for the verifiable presentation
  export const ZK_STATEMENTS = [
      {
      type: "RevealAttribute",
      attributeTag: "nationalIdNo",
      },
      {
      type: "RevealAttribute",
      attributeTag: "nationality",
      },
      {
      type: "AttributeInRange",
      attributeTag: "dob",
      lower: "18000101",
      upper: "20060802",  // Ensures user is over 18 years old
      },
      {
      type: "AttributeNotInSet",
      attributeTag: "countryOfResidence",
      set: ["US", "KP", "RU"],  // Not from USA, North Korea, or Russia
      },
  ] as AtomicStatementV2[];

  // The number of blocks after the `best block` (top of chain), where the `recent block` is located.

>   // The `recent block hash` is included in signatures and ZK proofs to ensure they expire.

Reviewer: Signatures seem not the focus of this tutorial.

```suggestion
  // The `recent block hash` is included in ZK proofs to ensure they expire.
```

>   // The `recent block hash` is included in signatures and ZK proofs to ensure they expire.

Reviewer: Worth to mention that the `verifier` needs to check that a recent `block_hash` was used in the ZK proof for proofs to expiry.

  export const RECENT_BLOCK_DURATION = 10n;

The ZK statements use Concordium's ``AtomicStatementV2`` type to define:

* **CONTEXT_STRING**: A domain separator that ensures ZK proofs are specific to this application.

* **RECENT_BLOCK_DURATION**: The number of blocks behind the best block to use for challenge generation (10 blocks in this case). This creates time-limited proofs.

* **ZK statements types**:

  1. **RevealAttribute**: The simplest statement type that reveals a specific attribute's value.

  * We request the user's ``nationalIdNo`` and ``nationality`` attributes.
  * These values will be directly visible to the application.

  2. **AttributeInRange**: A statement that proves an attribute falls within a specified range, without revealing the actual value.

  * We verify the ``dob`` (date of birth) falls between Jan 1, 1800, and Aug 2, 2006.
  * This effectively proves the user is over 18 years old on Aug 2, 2024.
  * The actual date of birth remains private.

  3. **AttributeNotInSet**: A statement that proves an attribute is not in a specified set of values.

>   * We verify the ``countryOfResidence`` is not in the set ["US", "KP", "RU"].

Reviewer: Worth mention with standard we follow for these country codes so people can use other countries easily.

Author: Note about the `ISO-3166-1 alpha-2` standard was added.

For security, ZK proofs need to be time-limited. We achieve this by including a recent block hash in the challenge:

.. code-block:: javascript

  // From utils.ts, imports ommited for brevity
  export async function getRecentBlock(
    grpcClient: ConcordiumGRPCClient | undefined,
  ): Promise<RecentBlock> {
    if (!grpcClient) {
      throw Error(`'grpcClient' is undefined`);
    }
    // Get the best block height (top of chain)
    const bestBlockHeight = (await grpcClient.client.getConsensusInfo(""))
      ?.response.bestBlockHeight;

    if (!bestBlockHeight) {
      throw Error(`Couldn't get 'bestBlockHeight' from chain`);
    }
    // Calculate recent block height (e.g., 10 blocks behind)
    const recentBlockHeight = bestBlockHeight.value - RECENT_BLOCK_DURATION;
    // Get the hash of the block at that height
    const recentBlockHash = (
      await grpcClient.getBlocksAtHeight(recentBlockHeight)
    )[0];

    if (!recentBlockHash) {
      throw Error(`Couldn't get 'recentBlockHash' from chain`);
    }

    return { blockHash: recentBlockHash, blockHeight: recentBlockHeight };
  }

This function retrieves a recent block to use in challenge generation:

> * **Security purpose**: Including a recent block hash in the challenge creates time-limited ZK proofs. After the chain progresses further, the proofs will no longer be valid, preventing `replay attacks <https://en.wikipedia.org/wiki/Replay_attack>`_.

Reviewer: Worth to mention that it is the `verifiers` responsibility to check that the `block_hash` included in the proof is recent so that proofs expiry. Proofs are not in itself valid/invalid (the verifier decides what proofs it accepts as valid/invalid).

Author: Done, line 121.

> * **Security purpose**: Including a recent block hash in the challenge creates time-limited ZK proofs. After the chain progresses further, the proofs will no longer be valid, preventing `replay attacks <https://en.wikipedia.org/wiki/Replay_attack>`_.

Reviewer: In general, linking to wikipedia should be avoided (not a good/reliable source since it can be added rather easily by random people).

Author: Wikipedia links were replaced or removed. This should be done in all the docs at one point.

* **Blockchain interaction**: The function uses ``ConcordiumGRPCClient`` to communicate with a Concordium node:

  1. First, it retrieves the current best block height using ``getConsensusInfo()``
  2. It calculates a "recent" block by subtracting ``RECENT_BLOCK_DURATION`` (10 blocks)
  3. It fetches the hash of that block using ``getBlocksAtHeight()``

* **Error handling**: The function includes checks to ensure valid data at each step:

  * Verifies the GRPC client is available
  * Checks that the best block height was successfully retrieved
  * Confirms a block hash was found at the calculated height

* **Return value**: The function returns both the block hash and height.

Generating and verifying ZK proofs
----------------------------------

Now let's implement the ZK proof generation and verification. First, we'll set up the component structure and initial state:

.. code-block:: typescript

  // From components/proof/Proof.tsx, imports ommited for brevity

  const Proof = () => {

>     const navigate = useNavigate();

Reviewer: ```suggestion

```

      issuers: [0, 1, 2, 3, 4, 5, 6, 7],
    },
    statement: ZK_STATEMENTS,
  };

This section creates a secure challenge for the ZK proof:

1. It combines the recent block hash with a context string (domain separator) into a buffer array
2. It uses SHA-256 to hash this combination, creating a unique, time-limited challenge
3. It defines a credential statement that specifies:

  * The acceptable identity providers (issuers 0-7)
  * The ZK statements defining what should be proved (imported from constants)

The context string ensures that proofs generated for one application cannot be reused in another application, improving security.

Now let's request the ZK proof from the wallet:

.. code-block:: typescript

  // 4. Request ZK proof from the wallet
  const presentation = await provider.requestVerifiablePresentation(
    challenge,
    [credentialStatement],
  );
  setIsLoading(false);
  // 5. Extract revealed attributes from the proof
  setIdNumber(
    presentation.verifiableCredential[0].credentialSubject.proof
      .proofValue[0].attribute,
  );
  setNationality(
    presentation.verifiableCredential[0].credentialSubject.proof
      .proofValue[1].attribute,
  );

>   // 6. Mark proof as valid (in production, verify on backend)

Reviewer: Worth to extremely highlight this fact.

```
  // 6. Mark proof as valid (this is unsafe, in production you have to verify the ZK proof at a backend before progressing)

```

Author: A warning was added after point 6, line 263 to highlight this.

1. It calls ``requestVerifiablePresentation()`` on the wallet provider

  * This prompts the wallet to show a UI for the user to approve the proof generation
  * The wallet handles all the complex cryptography to generate the proof

2. Once the proof is returned, it extracts revealed attributes:

  * The ID number from the first proof value
  * The nationality from the second proof value

3. It marks the proof as valid and exits the loading state

In a production environment, you would send the proof to a backend for cryptographic verification. In this demo, we're assuming any returned proof is valid for simplicity.
You can explore this `github repository <https://github.com/Concordium/concordium-web3id>`_ which contains a verifier.

The component's render function handles different UI states (loading, verification success, and error states) and displays the results of the ZK proof to the user.
The complete implementation of this component can be found in the example repository, in ``frontend/src/components/proof/Proof.tsx``

The most powerful aspect of this implementation is that the wallet handles all the complex cryptography. The dApp only needs to:

> #. Define what should be proved

Reviewer: ```suggestion
#. Define what should be proven
```

> #. Generate a secure challenge

Reviewer: What is a `secure`/`not-secure` challenge?
```suggestion
#. Generate a challenge
```

.. _zk-proofs-tutorial:

===================

> ZK proofs structure

Reviewer: The goal of the tutorial is to teach the reader more about zk proof generation (not only about the structure of zk proofs), so maybe we can reflect that in the header.
```suggestion
ZK proof generation
```

===================

Now we'll explore how to generate zero-knowledge proofs using the ID Layer, focusing on how ZK statements are structured.

Defining ZK statements
----------------------

First, let's look at the ZK statements defined in the ``constants`` file:

.. code-block:: javascript

  // From constants.ts

  import { AtomicStatementV2 } from "@concordium/web-sdk";

>   // Context string for signing messages and generating ZK proofs

Reviewer: ```suggestion
  // Context string for and generating ZK proofs
```

      .proofValue[1].attribute,
  );

  // 6. Mark proof as valid (in production, verify on backend)
  setValidZKProof(true);

.. warning::

  In a production environment, this proof **MUST** be verified on a secure backend!
  This frontend-only implementation is for **demonstration purposes only**.
  Proper implementation would send the presentation to a backend `verifier <https://github.com/Concordium/concordium-web3id/tree/main/services/web3id-verifier>`_.

This section handles the proof request and processing:

1. It calls ``requestVerifiablePresentation()`` on the wallet provider

  * This prompts the wallet to show a UI for the user to approve the proof generation
  * The wallet handles all the complex cryptography to generate the proof

2. Once the proof is returned, it extracts revealed attributes:

  * The ID number from the first proof value
  * The nationality from the second proof value

3. It marks the proof as valid and exits the loading state

In a production environment, you would send the proof to a backend for cryptographic verification. In this demo, we're assuming any returned proof is valid for simplicity.

> You can explore this `github repository <https://github.com/Concordium/concordium-web3id>`_ which contains a verifier.

Reviewer: ```suggestion
You can explore this `github repository <https://github.com/Concordium/concordium-web3id/tree/main/services/web3id-verifier>`_ which contains a verifier.
```
